### ISSUE TITLE
AWS SSM ParameterStore provider requests full access

### ISSUE THREAD

**Issue by dzavalkin-scayle** [external-secrets/external-secrets#1839]
[Documentation](https://external-secrets.io/v0.7.0/provider/aws-parameter-store/#iam-policy) says it is possible to provide access to certain path in AWS SSM ParameterStore:

```hcl
"Resource": "arn:aws:ssm:us-east-2:1234567889911:parameter/dev-*"
```

Unfortunately, in reality current 0.7.0 refuses to work without full access, i.e. `arn:aws:ssm:us-east-2:1234567889911:`:

```
Events:
  Type     Reason        Age                   From              Message
  ----     ------        ----                  ----              -------
  Warning  UpdateFailed  88s (x111 over 108m)  external-secrets  (combined from similar events): AccessDeniedException: User: arn:aws:sts::1234567889911:assumed-role/poc-service-role/external-secrets-provider-aws is not authorized to perform: ssm:DescribeParameters on resource: arn:aws:ssm:us-east-2:1234567889911:* because no identity-based policy allows the ssm:DescribeParameters action
           status code: 400, request id: a57d237e-2dcd-4d60-b034-7b32cf58d645
```

In addition to that, provided IAM policy 

```
      "Action": [
        "ssm:GetParameterWithContext",
        "ssm:ListTagsForResourceWithContext",
        "ssm:DescribeParametersWithContext",
      ],
 ```

is totally invalid - there are no such permissions:
![image_iam_policy_error](https://user-images.githubusercontent.com/113674675/209129936-4b3a9d02-b15e-45db-9bec-44c2f78d9e19.png)

**Comment by gusfcarvalho** [comment]
Hi @dzavalkin-scayle ! This is because you are using dataFrom.find. The underlying AWS call needs to have these open permissions in order to run it.

**Comment by martina-if** [comment]
Hi @gusfcarvalho , I'm seeing the same issue while migrating from KES. Is there a way to still have the policy limited to a specific subtree of parameter store keys while retrieving all keys in the subtree via a path?

**Comment by moolen** [comment]
API-wise this seems possible. Though this would be a breaking change, because it is a different API call that requires different permissions.

This is problematic :thinking:. I remember there was another issue related to this problem, but can't find it right now.

https://docs.aws.amazon.com/sdk-for-go/api/service/ssm/#SSM.GetParametersByPath

**Comment by martina-if** [comment]
Hi @moolen , I have been using a configuration like this in the `externalsecret`:

```
  dataFrom:
    - find:
        path: /abc/def/
        name:
          regexp: ".*"
```
But even when giving ESO access to all SSM parameters I get:

```
AccessDeniedException: The ciphertext refers to a customer master key that does not exist, does not exist in this region, or you are not allowed to access. (Service: AWSKMS; Status Code: 400; Error Code: AccessDeniedException; ...
``` 

I _think_ I also gave it access to all kms keys but I may have done that part wrong.

Is this lookup by path supported for AWS SSM? And if it is, I think it's a big limitation that one needs to give it access to basically all parameters in that account since it goes against the principle of less privilege 😕

**Comment by martina-if** [comment]
I had some more time to try things out with a fresh mind and saw that it just needs to describe these secrets but not necessarily decrypt them so I think that should be good enough. Might be worth adding a note about the different permissions to the kes-to-eso migration guide. I can open a PR if people think that's a good idea.

**Comment by moolen** [comment]
> Otherwise it makes it difficult to apply the principle of least privilege :confused:

100% this! I'd be happy to support `GetParametersByPath`. Though we should make this a non-breaking change by keeping the same behavior as a fallback.

I made this issue a `kind/feature`. If anyone reading this wants to contribute feel free to do so ;)

tl;dr What we want:

* use `GetParametersByPath` API calls to implement `dataFrom[].find` 
* keep current behavior of `dataFrom[].find` as a fallback
* add docs about the different permissions required for using this 

> Might be worth adding a note about the different permissions to the kes-to-eso migration guide

TBH i don't know how many people do use it and this issue is very specific. I don't really see it's worth the effort. But feel free if you want to of course!
